src/dune_engine/process.ml

```diff
@@ -601,36 +603,51 @@ let run_internal ?dir ?(stdout_to = Io.stdout) ?(stderr_to = Io.stderr)
       | None, None -> ()
       | _, _ -> assert false);
       Option.iter response_file ~f:Path.unlink;
+      let actual_stdout =
+        match stdout_capture with
+        | `No_capture -> lazy ""
+        | `Capture fn -> lazy (Stdune.Io.read_file fn)
+      in
+      let actual_stderr =
+        match stderr_capture with
+        | `No_capture
+        | `Merged_with_stdout ->
+          lazy ""
+        | `Capture fn -> lazy (Stdune.Io.read_file fn)
+      in
+      let outputs_are_as_expected =
+        (stdout_on_success <> Must_be_empty || Lazy.force actual_stdout = "")
+        && (stderr_on_success <> Must_be_empty || Lazy.force actual_stderr = "")
+      in
       let exit_status' : Exit_status.t =
         match exit_status with
-        | WEXITED n when ok_codes n -> Ok n
+        | WEXITED n when outputs_are_as_expected && ok_codes n -> Ok n
         | WEXITED n -> Error (Failed n)
         | WSIGNALED n -> Error (Signaled (Signal.name n))
         | WSTOPPED _ -> assert false
       in
       let success = Result.is_ok exit_status' in
-      let read_and_destroy fn ~swallow_on_success =
+      let read_and_destroy fn actual_output
```

Comment:
Got a bit confused here, perhaps because of different naming. I would call this function `swallow_on_success_if_requested` to keep naming consistent.

Reply:
seems fine